File: backend/app/routers/clients.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from uuid import UUID

from app.core.database import get_db
from app.models.user import User
from app.models.client import Client
from app.schemas.client import ClientCreate, ClientUpdate, ClientResponse
from app.dependencies.auth import get_current_user, require_admin

router = APIRouter(prefix="/clients", tags=["Clientes"])


@router.get("/", response_model=List[ClientResponse])
def list_clients(
    active_only: bool = True,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    q = db.query(Client)
    if active_only:
        q = q.filter(Client.active == True)
    return q.order_by(Client.name).all()


@router.post("/", response_model=ClientResponse, status_code=201)
def create_client(
    body: ClientCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    client = Client(**body.model_dump())
    db.add(client)
    db.commit()
    db.refresh(client)
    return client


@router.get("/{client_id}", response_model=ClientResponse)
def get_client(
    client_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    client = db.query(Client).filter(Client.id == client_id).first()
    if not client:
        raise HTTPException(status_code=404, detail="Cliente no encontrado.")
    return client


@router.put("/{client_id}", response_model=ClientResponse)
def update_client(
    client_id: UUID,
    body: ClientUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    client = db.query(Client).filter(Client.id == client_id).first()
    if not client:
        raise HTTPException(status_code=404, detail="Cliente no encontrado.")
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(client, field, value)
    db.commit()
    db.refresh(client)
    return client


@router.delete("/{client_id}", status_code=204)
def delete_client(
    client_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    client = db.query(Client).filter(Client.id == client_id).first()
    if not client:
        raise HTTPException(status_code=404, detail="Cliente no encontrado.")
    db.delete(client)
    db.commit()

File: frontend/src/api/clients.ts
import { apiClient } from './client'
import type { Client } from '../types'

export interface ClientCreate {
  name: string
  rut?: string
  phone?: string
  email?: string
  address?: string
  notes?: string
}

export interface ClientUpdate extends Partial<ClientCreate> {
  active?: boolean
}

export const clientsApi = {
  list: async (activeOnly = true): Promise<Client[]> => {
    const { data } = await apiClient.get<Client[]>('/clients/', {
      params: { active_only: activeOnly },
    })
    return data
  },

  get: async (id: string): Promise<Client> => {
    const { data } = await apiClient.get<Client>(`/clients/${id}`)
    return data
  },

  create: async (body: ClientCreate): Promise<Client> => {
    const { data } = await apiClient.post<Client>('/clients/', body)
    return data
  },

  update: async (id: string, body: ClientUpdate): Promise<Client> => {
    const { data } = await apiClient.put<Client>(`/clients/${id}`, body)
    return data
  },

  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/clients/${id}`)
  },
}
